### ISSUE TITLE
Scylla driver 0.2.1 and 0.3.0 more than twice slower than Datastax driver with cassandra_cpp bindings

### ISSUE THREAD

**Issue by pkolaczk** [scylladb/scylla-rust-driver#338]
Having seen your benchmarks, an idea has occurred to me to switch my benchmarking tool latte (github.com/pkolaczk/latte) from cassandra_cpp to scylla to gain even more performance. To my dissapointment, it turned out that caused a performance regression :(

Running the same write workload against single local node of Cassandra 4.0.1 (snapshot version I had lying around), well warmed-up instance.

cassandra_cpp (master branch of latte):
```
$ latte run write -t 2

CONFIG =================================================================================================
            Date        : Sun, 14 Nov 2021                                                         
            Time        : 09:57:33 +0100                                                           
             Tag        :                                                                          
        Workload        : write                                                                    
      Compaction        : STCS                                                                     
      Partitions        :   1000000                                                                 
         Columns        :         1                                                                 
     Column size     [B]:        16                                                                 
         Threads        :         2                                                                 
     Connections        :         1                                                                 
 Max parallelism   [req]:       256                                                                 
        Max rate [req/s]:                                                                          
          Warmup   [req]:         1                                                                 
      Iterations   [req]:   1000000                                                                 
        Sampling     [s]:       1.0                                                                 

LOG ====================================================================================================
    Time  Throughput        ----------------------- Response times [ms]---------------------------------
     [s]     [req/s]           Min        25        50        75        90        95        99       Max
   0.000      107061          0.33      2.12      3.01      4.97      8.96     13.39     20.05     36.16
   1.000      179131          0.22      1.87      2.41      2.99      3.62      4.14      6.11     14.83
   2.000      174108          0.18      1.88      2.42      3.03      3.72      4.26      6.10     18.70
   3.000      160531          0.20      1.80      2.37      3.03      3.76      4.26      7.93    102.97
   4.000      179502          0.24      1.81      2.39      2.99      3.60      4.06      5.83     18.14
   5.000      169947          0.26      1.87      2.45      3.09      3.83      4.47      6.28     22.75
   6.000      166527          0.18      1.92      2.45      3.18      4.00      4.58      6.49      9.57

SUMMARY STATS ==========================================================================================
         Elapsed     [s]:     6.179                                                                 
        CPU time     [s]:     9.334          ( 18.9%)                                               
       Completed   [req]:   1000000          (100.0%)                                               
          Errors   [req]:         0          (  0.0%)                                               
      Partitions        :   1000000                                                                 
            Rows        :   1000000                                                                 
         Samples        :         7                                                                 
Mean sample size   [req]:    142857                                                                 
     Parallelism   [req]:     196.4          ( 76.7%)                                               
      Throughput [req/s]:    161852 ± 33268                                                         
 Mean resp. time    [ms]:      2.79 ± 0.76                                                          
```

scylla driver 0.3.0 (scylla-driver branch):

```
$ latte run write -t 2

CONFIG =================================================================================================
            Date        : Sun, 14 Nov 2021                                                         
            Time        : 09:59:47 +0100                                                           
             Tag        :                                                                          
        Workload        : write                                                                    
      Compaction        : STCS                                                                     
      Partitions        :   1000000                                                                 
         Columns        :         1                                                                 
     Column size     [B]:        16                                                                 
         Threads        :         2                                                                 
     Connections        :         1                                                                 
 Max parallelism   [req]:       256                                                                 
        Max rate [req/s]:                                                                          
          Warmup   [req]:         1                                                                 
      Iterations   [req]:   1000000                                                                 
        Sampling     [s]:       1.0                                                                 

LOG ====================================================================================================
    Time  Throughput        ----------------------- Response times [ms]---------------------------------
     [s]     [req/s]           Min        25        50        75        90        95        99       Max
   0.000       87208          1.38      4.87      5.28      5.77      6.39      6.85      9.18     77.31
   1.000       91906          1.51      4.96      5.38      5.91      6.53      6.98      8.50     18.83
   2.000       90949          1.40      5.04      5.47      5.95      6.50      6.92      7.95     11.20
   3.000       90669          1.04      5.03      5.45      5.99      6.71      7.20      8.89     11.24
   4.000       89914          1.04      5.08      5.46      6.00      6.68      7.31      8.46     19.90
   5.000       91066          1.39      4.98      5.42      5.91      6.61      7.35      8.87     12.80
   6.000       87022          1.29      5.11      5.66      6.33      7.19      7.81      9.38     13.32
   7.000       88806          2.03      5.10      5.56      6.14      6.76      7.27      8.67     12.85
   8.000       79710          1.49      5.16      5.55      6.00      6.58      7.39     10.34    114.88
   9.000       89626          1.37      5.09      5.52      6.01      6.60      7.26      9.36     19.09
  10.000       88613          1.52      5.18      5.55      6.08      6.76      7.35      8.65     15.78
  11.000       84421          1.41      5.34      5.95      6.66      7.65      8.30      9.77     13.86

SUMMARY STATS ==========================================================================================
         Elapsed     [s]:    11.291                                                                 
        CPU time     [s]:    18.469          ( 20.4%)                                               
       Completed   [req]:   1000000          (100.0%)                                               
          Errors   [req]:         0          (  0.0%)                                               
      Partitions        :   1000000                                                                 
            Rows        :   1000000                                                                 
         Samples        :        12                                                                 
Mean sample size   [req]:     83333                                                                 
     Parallelism   [req]:     239.0          ( 93.4%)                                               
      Throughput [req/s]:     88571 ± 3637                                                          
 Mean resp. time    [ms]:      5.69 ± 0.24                                                          
```

See in particular this - this is client CPU time:
```
CPU time     [s]:     9.334          ( 18.9%)                         // cassandra_cpp
CPU time     [s]:    18.469          ( 20.4%)                        // scylla driver
```

I can see a similar difference when running only 1 tokio thread.
Also Scylla driver 0.2.1 wasn't any better.

I took flamegraphs of both runs - maybe that would help you figure out what I'm doing wrong.
[flamegraphs.zip](https://github.com/scylladb/scylla-rust-driver/files/7533446/flamegraphs.zip)

One immediate difference is that scylla seems to be spending a lot more time in write. 

BTW: I really like the API of scylla driver much better and the fact this is a native Rust driver so a lot easier to install and use than a hybrid of rust bindings + cpp libs, but performance is a bummer at the moment. 

Operating system: Ubuntu 21.10, kernel `5.13.0-21-generic #21-Ubuntu SMP Tue Oct 19 08:59:28 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`

**Comment by psarna** [comment]
/cc @piodul @cvybhu @havaker 

Thanks for reporting, we have an ongoing effort of benchmarking and profiling the driver under various workloads at the moment in order to reduce the bottlenecks, your flamegraphs would definitely help (I haven't peeked at them yet but will definitely do so on Monday). We have also already revealed one performance problem with latency metrics (https://github.com/scylladb/scylla-rust-driver/issues/330) which will be taken care of soon.

**Comment by pkolaczk** [comment]
I looked at the syscall stats, and Scylla driver is doing one `sendto` and 2 `recvfrom` per query.

Scylla stats:
```
         1 000 014      syscalls:sys_enter_sendto                                   
         2 020 118      syscalls:sys_enter_recvfrom 
            82 638      syscalls:sys_enter_epoll_wait                                   
                19      syscalls:sys_enter_read                                     
            38 590      syscalls:sys_enter_write                                    
                 0      syscalls:sys_enter_writev                                   
            34 637      syscalls:sys_enter_futex                                                
           394 200      syscalls:sys_enter_sched_yield                                   
```            
cassandra_cpp stats:

```
                 6      syscalls:sys_enter_sendto                                   
                 0      syscalls:sys_enter_recvfrom     
            30 712      syscalls:sys_enter_epoll_wait    
            42 315      syscalls:sys_enter_read                                     
             4 414      syscalls:sys_enter_write 
            13 290      syscalls:sys_enter_writev                                   
           391 663      syscalls:sys_enter_futex                                    
            48 723      syscalls:sys_enter_sched_yield                                   
            30 712      syscalls:sys_enter_rt_sigtimedwait    
```

**Comment by havaker** [comment]
I guess this corresponds to the way we read incoming frames. Header is read first, then body.
https://github.com/scylladb/scylla-rust-driver/blob/75345fcf7dc6de601f8dfcfa2adb1bbd425b591a/scylla/src/frame/mod.rs#L104
https://github.com/scylladb/scylla-rust-driver/blob/75345fcf7dc6de601f8dfcfa2adb1bbd425b591a/scylla/src/frame/mod.rs#L133
As we do not use any buffered io (`read_exact` and `read_buf` are called on raw `TcpStream` when not using ssl), this possibly translates to a double `recvfrom`.

**Comment by pkolaczk** [comment]
With async requests you could probably also buffer writes and send more than one query with a single `write/sendto`. At least this is what the c++ driver is doing.

**Comment by psarna** [comment]
@pkolaczk could you provide us as much context about your testing environment as possible? Number of nodes in the test cluster, how big the machines are (in terms of cpu and ram), are they in a local network or is it a mixed environment (e.g. client is local and nodes are on AWS), etc. I'm asking because I reran the latte command on our local test cluster - which resides on the same local network as `latte`, but on a different machine - and the results are vastly different, so I'm probably just comparing apples to oranges, which makes no sense, and instead it would be much better to replicate your environment as close as possible. I ran latte against both a Scylla node and a Cassandra 4.0.1 node and both with cassandra-cpp and scylla driver (so 4 runs total) and my results tell a different story - full output of all results is pasted at the bottom of this post:

1. `scylla` driver ran against a Scylla node has 2x worse tail latency (it also has better minimum latency, but it's much less important to us than tail latency), which is concerning and should be fixed, but throughput characteristics and latencies up to 95th percentile are quite comparable to the `cassandra-cpp` driver.
2. `scylla` driver ran against a Cassandra 4.0.1 node seems to have reduced throughput compared to `cassandra-cpp`, but it's an order of 10-15%, not 50%. Tail latencies max up to 40ms for both drivers.

Comand used for all runs: `latte run write -t 2`

### scylla 4.5 node with 2 shards; driver=scylla:
```
CONFIG =================================================================================================
            Date        : Mon, 15 Nov 2021                                                         
            Time        : 09:39:31 +0100                                                           
             Tag        :                                                                          
        Workload        : write                                                                    
      Compaction        : STCS                                                                     
      Partitions        :   1000000                                                                 
         Columns        :         1                                                                 
     Column size     [B]:        16                                                                 
         Threads        :         2                                                                 
     Connections        :         1                                                                 
 Max parallelism   [req]:       256                                                                 
        Max rate [req/s]:                                                                          
          Warmup   [req]:         1                                                                 
      Iterations   [req]:   1000000                                                                 
        Sampling     [s]:       1.0                                                                 

LOG ====================================================================================================
    Time  Throughput        ----------------------- Response times [ms]---------------------------------
     [s]     [req/s]           Min        25        50        75        90        95        99       Max
   0.000      163632          0.23      2.14      2.68      3.36      4.23      4.91      6.51     10.87
   1.000      165229          0.53      2.15      2.68      3.35      4.13      4.69      5.98     10.94
   2.000      165927          0.57      2.15      2.67      3.32      4.09      4.66      5.84     10.47
   3.000      163633          0.69      2.18      2.70      3.37      4.18      4.76      6.06     10.56
   4.000      164118          0.72      2.17      2.69      3.36      4.18      4.76      6.04     10.92
   5.000      169816          0.57      2.11      2.63      3.26      4.04      4.64      5.84     10.10
   6.000      179407          0.96      2.12      2.56      3.10      3.69      4.06      5.04      8.41

SUMMARY STATS ==========================================================================================
         Elapsed     [s]:     6.043                                                                 
        CPU time     [s]:    17.401          (  9.0%)                                               
       Completed   [req]:   1000000          (100.0%)                                               
          Errors   [req]:         0          (  0.0%)                                               
      Partitions        :   1000000                                                                 
            Rows        :   1000000                                                                 
         Samples        :         7                                                                 
Mean sample size   [req]:    142857                                                                 
     Parallelism   [req]:     233.5          ( 91.2%)                                               
      Throughput [req/s]:    165491 ± 5252                                                          
 Mean resp. time    [ms]:      2.84 ± 0.07                                                          

THROUGHPUT DISTRIBUTION [req/s] ========================================================================
             Min        :    163632                                                                 
               1        :    163632                                                                 
               2        :    163632                                                                 
               5        :    163632                                                                 
              10        :    163632                                                                 
              25        :    163633                                                                 
              50        :    165229                                                                 
              75        :    169816                                                                 
              90        :    179406                                                                 
              95        :    179406                                                                 
             Max        :    179406                                                                 

MEAN RESPONSE TIMES [ms] ===============================================================================
             Min        :      0.56 ± 0.29                                                          
              25        :      2.15 ± 0.03                                                          
              50        :      2.67 ± 0.05                                                          
              75        :      3.33 ± 0.09                                                          
              90        :      4.14 ± 0.13                                                          
              95        :      4.73 ± 0.18                                                          
              98        :      5.49 ± 0.33                                                          
              99        :      6.04 ± 0.39                                                          
            99.9        :      7.65 ± 0.45                                                          
           99.99        :      9.29 ± 0.54                                                          
             Max        :     10.62 ± 0.74                                                          

RESPONSE TIME DISTRIBUTION =============================================================================
Percentile    Resp. time [ms]  ------------------------------- Count -----------------------------------
   0.00000          0.23               0   
   0.00110          0.28              11   
   0.00170          0.35               6   
   0.00320          0.44              15   
   0.00560          0.56              24   
   0.01700          0.69             114   
   0.05410          0.87             371   
   0.31540          1.09            2613   
   1.83680          1.36           15214   *
   8.09970          1.70           62629   ******
  23.79350          2.12          156938   ***************
  49.03740          2.65          252439   *************************
  74.56750          3.32          255301   *************************
  90.10900          4.15          155415   ***************
  97.09600          5.18           69870   ******
  99.41460          6.48           23186   **
  99.94690          8.10            5323   
  99.99760         10.12             507   
 100.00000         12.65              24   
```

### scylla 4.5 node with 2 shards; driver=cassandra-cpp:
```
CONFIG =================================================================================================
            Date        : Mon, 15 Nov 2021                                                         
            Time        : 09:42:32 +0100                                                           
             Tag        :                                                                          
        Workload        : write                                                                    
      Compaction        : STCS                                                                     
      Partitions        :   1000000                                                                 
         Columns        :         1                                                                 
     Column size     [B]:        16                                                                 
         Threads        :         2                                                                 
     Connections        :         1                                                                 
 Max parallelism   [req]:       256                                                                 
        Max rate [req/s]:                                                                          
          Warmup   [req]:         1                                                                 
      Iterations   [req]:   1000000                                                                 
        Sampling     [s]:       1.0                                                                 

LOG ====================================================================================================
    Time  Throughput        ----------------------- Response times [ms]---------------------------------
     [s]     [req/s]           Min        25        50        75        90        95        99       Max
   0.000      169563          0.73      2.47      2.76      3.06      3.34      3.53      4.04      5.74
   1.000      166879          0.85      2.55      2.85      3.12      3.35      3.50      3.76      4.67
   2.000      165945          0.88      2.65      2.94      3.19      3.42      3.57      3.85      5.33
   3.000      164219          0.60      2.76      3.00      3.21      3.43      3.57      3.84      5.45
   4.000      164712          1.63      2.77      2.99      3.20      3.42      3.55      3.79      4.94
   5.000      164974          1.60      2.77      2.99      3.20      3.42      3.54      3.76      5.12
   6.000      166402          2.05      2.82      3.04      3.24      3.41      3.51      3.70      3.88

SUMMARY STATS ==========================================================================================
         Elapsed     [s]:     6.023                                                                 
        CPU time     [s]:    21.671          ( 11.2%)                                               
       Completed   [req]:   1000000          (100.0%)                                               
          Errors   [req]:         0          (  0.0%)                                               
      Partitions        :   1000000                                                                 
            Rows        :   1000000                                                                 
         Samples        :         7                                                                 
Mean sample size   [req]:    142857                                                                 
     Parallelism   [req]:     240.2          ( 93.8%)                                               
      Throughput [req/s]:    166050 ± 2774                                                          
 Mean resp. time    [ms]:      2.91 ± 0.18                                                          

THROUGHPUT DISTRIBUTION [req/s] ========================================================================
             Min        :    164219                                                                 
               1        :    164219                                                                 
               2        :    164219                                                                 
               5        :    164219                                                                 
              10        :    164219                                                                 
              25        :    164712                                                                 
              50        :    165945                                                                 
              75        :    166879                                                                 
              90        :    169563                                                                 
              95        :    169563                                                                 
             Max        :    169563                                                                 

MEAN RESPONSE TIMES [ms] ===============================================================================
             Min        :      1.05 ± 0.86                                                          
              25        :      2.66 ± 0.24                                                          
              50        :      2.92 ± 0.17                                                          
              75        :      3.16 ± 0.11                                                          
              90        :      3.40 ± 0.06                                                          
              95        :      3.54 ± 0.03                                                          
              98        :      3.71 ± 0.06                                                          
              99        :      3.84 ± 0.14                                                          
            99.9        :      4.43 ± 0.39                                                          
           99.99        :      4.96 ± 0.43                                                          
             Max        :      5.20 ± 0.48                                                          

RESPONSE TIME DISTRIBUTION =============================================================================
Percentile    Resp. time [ms]  ------------------------------- Count -----------------------------------
   0.00000          0.60               0   
   0.00030          0.75               3   
   0.00360          0.93              33   
   0.06700          1.17             634   
   0.29500          1.46            2280   
   1.12280          1.82            8278   
   6.03560          2.28           49128   ****
  41.73720          2.85          357016   ***********************************
  95.51680          3.56          537796   *****************************************************
  99.86860          4.45           43518   ****
  99.99990          5.57            1313   
 100.00000          6.96               1   
```

### Cassandra 4.0.1 node; driver=scylla:
```
CONFIG =================================================================================================
            Date        : Mon, 15 Nov 2021                                                         
            Time        : 10:09:59 +0100                                                           
             Tag        :                                                                          
        Workload        : write                                                                    
      Compaction        : STCS                                                                     
      Partitions        :   1000000                                                                 
         Columns        :         1                                                                 
     Column size     [B]:        16                                                                 
         Threads        :         2                                                                 
     Connections        :         1                                                                 
 Max parallelism   [req]:       256                                                                 
        Max rate [req/s]:                                                                          
          Warmup   [req]:         1                                                                 
      Iterations   [req]:   1000000                                                                 
        Sampling     [s]:       1.0                                                                 

LOG ====================================================================================================
    Time  Throughput        ----------------------- Response times [ms]---------------------------------
     [s]     [req/s]           Min        25        50        75        90        95        99       Max
   0.000      165443          0.82      2.30      2.74      3.28      3.92      4.35      5.30      8.15
   1.000      147692          1.00      2.54      3.01      3.57      4.21      4.66      5.85     37.63
   2.000      141818          0.95      2.66      3.16      3.76      4.44      4.90      5.89     34.97
   3.000      142463          0.94      2.72      3.15      3.72      4.37      4.83      5.99     38.62
   4.000      137815          1.01      2.71      3.21      3.82      4.51      5.03      6.26     38.05
   5.000      163375          0.95      2.32      2.77      3.34      3.97      4.47      5.57      9.53
   6.000      158944          0.86      2.28      2.73      3.27      3.92      4.40      5.77     38.24

SUMMARY STATS ==========================================================================================
         Elapsed     [s]:     6.639                                                                 
        CPU time     [s]:    18.003          (  8.5%)                                               
       Completed   [req]:   1000000          (100.0%)                                               
          Errors   [req]:         0          (  0.0%)                                               
      Partitions        :   1000000                                                                 
            Rows        :   1000000                                                                 
         Samples        :         7                                                                 
Mean sample size   [req]:    142857                                                                 
     Parallelism   [req]:     233.2          ( 91.1%)                                               
      Throughput [req/s]:    150649 ± 14132                                                         
 Mean resp. time    [ms]:      3.14 ± 0.30                                                          

THROUGHPUT DISTRIBUTION [req/s] ========================================================================
             Min        :    137815                                                                 
               1        :    137815                                                                 
               2        :    137815                                                                 
               5        :    137815                                                                 
              10        :    137815                                                                 
              25        :    141818                                                                 
              50        :    147692                                                                 
              75        :    163375                                                                 
              90        :    165443                                                                 
              95        :    165443                                                                 
             Max        :    165443                                                                 

MEAN RESPONSE TIMES [ms] ===============================================================================
             Min        :      0.93 ± 0.08                                                          
              25        :      2.51 ± 0.25                                                          
              50        :      2.97 ± 0.26                                                          
              75        :      3.54 ± 0.29                                                          
              90        :      4.19 ± 0.31                                                          
              95        :      4.66 ± 0.33                                                          
              98        :      5.28 ± 0.36                                                          
              99        :      5.79 ± 0.39                                                          
            99.9        :     21.54 ± 18.67                                                         
           99.99        :     23.21 ± 19.34                                                         
             Max        :     28.05 ± 18.02                                                         

RESPONSE TIME DISTRIBUTION =============================================================================
Percentile    Resp. time [ms]  ------------------------------- Count -----------------------------------
   0.00000          0.82               0   
   0.01490          1.03             149   
   0.20890          1.29            1940   
   1.62670          1.61           14178   *
   7.89320          2.01           62665   ******
  26.41410          2.51          185209   ******************
  58.78590          3.14          323718   ********************************
  84.86030          3.93          260744   **************************
  96.37560          4.91          115153   ***********
  99.36790          6.14           29923   **
  99.77010          7.67            4022   
  99.79500          9.59             249   
  99.79520         11.99               2   
  99.79520         14.99               0   
  99.79520         18.73               0   
  99.79520         23.42               0   
  99.79520         29.27               0   
  99.99070         36.59            1955   
 100.00000         45.74              93 
```

### Cassandra 4.0.1 node; driver=cassandra-cpp:
```
CONFIG =================================================================================================
            Date        : Mon, 15 Nov 2021                                                         
            Time        : 09:57:31 +0100                                                           
             Tag        :                                                                          
        Workload        : write                                                                    
      Compaction        : STCS                                                                     
      Partitions        :   1000000                                                                 
         Columns        :         1                                                                 
     Column size     [B]:        16                                                                 
         Threads        :         2                                                                 
     Connections        :         1                                                                 
 Max parallelism   [req]:       256                                                                 
        Max rate [req/s]:                                                                          
          Warmup   [req]:         1                                                                 
      Iterations   [req]:   1000000                                                                 
        Sampling     [s]:       1.0                                                                 

LOG ====================================================================================================
    Time  Throughput        ----------------------- Response times [ms]---------------------------------
     [s]     [req/s]           Min        25        50        75        90        95        99       Max
   0.000      148053          0.59      2.44      2.68      2.95      3.89      5.93     13.62     41.34
   1.000      176020          0.84      2.46      2.68      2.83      2.96      3.04      3.30     33.79
   2.000      175408          0.68      2.55      2.73      2.85      2.97      3.05      3.32     31.26
   3.000      178234          0.71      2.48      2.73      2.92      3.03      3.10      3.30      4.11
   4.000      168812          1.10      2.68      2.82      2.95      3.06      3.13      3.37     33.53
   5.000      168269          1.28      2.68      2.83      2.95      3.05      3.12      3.42     34.53

SUMMARY STATS ==========================================================================================
         Elapsed     [s]:     5.913                                                                 
        CPU time     [s]:    20.477          ( 10.8%)                                               
       Completed   [req]:   1000000          (100.0%)                                               
          Errors   [req]:         0          (  0.0%)                                               
      Partitions        :   1000000                                                                 
            Rows        :   1000000                                                                 
         Samples        :         6                                                                 
Mean sample size   [req]:    166667                                                                 
     Parallelism   [req]:     238.5          ( 93.2%)                                               
      Throughput [req/s]:    169146 ± 14887                                                         
 Mean resp. time    [ms]:      2.84 ± 0.27                                                          

THROUGHPUT DISTRIBUTION [req/s] ========================================================================
             Min        :    148053                                                                 
               1        :    148053                                                                 
               2        :    148053                                                                 
               5        :    148053                                                                 
              10        :    148053                                                                 
              25        :    168269                                                                 
              50        :    168812                                                                 
              75        :    176020                                                                 
              90        :    178234                                                                 
              95        :    178234                                                                 
             Max        :    178234                                                                 

MEAN RESPONSE TIMES [ms] ===============================================================================
             Min        :      0.87 ± 0.40                                                          
              25        :      2.55 ± 0.17                                                          
              50        :      2.74 ± 0.12                                                          
              75        :      2.91 ± 0.07                                                          
              90        :      3.14 ± 0.46                                                          
              95        :      3.51 ± 1.47                                                          
              98        :      4.24 ± 3.62                                                          
              99        :      4.86 ± 5.33                                                          
            99.9        :     27.50 ± 16.71                                                         
           99.99        :     28.39 ± 17.41                                                         
             Max        :     29.25 ± 17.64                                                         

RESPONSE TIME DISTRIBUTION =============================================================================
Percentile    Resp. time [ms]  ------------------------------- Count -----------------------------------
   0.00000          0.59               0   
   0.00270          0.74              27   
   0.02410          0.92             214   
   0.11260          1.15             885   
   0.49590          1.44            3833   
   1.82740          1.80           13315   *
   8.39230          2.25           65649   ******
  60.79330          2.82          524010   ****************************************************
  97.82530          3.52          370320   *************************************
  98.56210          4.40            7368   
  98.93050          5.50            3684   
  99.21480          6.88            2843   
  99.36460          8.60            1498   
  99.55370         10.75            1891   
  99.64610         13.44             924   
  99.69270         16.80             466   
  99.69280         21.00               1   
  99.69280         26.24               0   
  99.93180         32.81            2390   
  99.99590         41.01             641   
 100.00000         51.26              41   
```

P.S. Huge thanks for coding `latte` and porting it to scylla-rust-driver! Such a tool with both `cassandra-cpp` and `scylla` backends is priceless for profiling and investigating performance bugs.

**Comment by psarna** [comment]
Oh and CPU time in my runs was mildly in favor of `scylla` over `cassandra-cpp` too, in both cases. By the way - could you please post the output of `cargo metadata` from latte? It would also help to run on matching versions of all the dependencies, e.g. I have tokio 1.13 locally, but both latte and scylla are compatible with lower versions too, starting with 1.1.

**Comment by psarna** [comment]
I just finished running both test cases against a cluster on Scylla Cloud, when @piodul pointed out that you indeed described the environment in the description, mentioning a "single local node". Does "local" mean LAN or the very same machine? We tend to avoid running benchmarks locally due to several points that make the results rather unreliable:
1. the driver and the node compete for resources locally with each other
2. the driver and the node compete with other processes active on your system, assuming that it's not a dedicated machine picked for the test
3. all the networking goes through loopback, which does not give realistic results compared to how distributed databases are usually used

We'd still be very much interested in reproducing the differences you achieved locally, but could you also consider running latte against a non-local cluster? I just finished my experiments with a 3-node cluster running on AWS and both drivers reached nearly identical throughput and latencies.

**Comment by psarna** [comment]
I reran the tests in a fully local environment, i.e. both the node and the driver residing on the same machine, and the conclusion is as follows - `scylla` has a cap for **minimal** latency on a single connection which is higher than the one in `cassandra-cpp`, which translates to reduced throughput in a single connection test **and** a hyper-fast network, of which loopback is the example. I did reproduce your result - around ~60kops for `scylla` and ~100kops for `cassandra-cpp` for a single connection, but then I reran the very same test with 2 and 4 conections per host, and the results are below:
1. `cassandra-cpp`: 2 connections: ~140kops; 4 connections: ~140kops
2. `scylla`: 2 connections: ~110kops; 4 connections: ~140kops 

So we indeed have latency issues, both with minimal and head latency, which we need to fix, but the minimal latency cap is only observable on really fast networks - I wasn't able to reproduce the results even on LAN, where both of the machines are only 1 switch apart - I've only seen the discrepancy on loopback. Also, starting with ~4 connections, the issue disappears even for loopback.

That also explains why I saw the opposite effect when testing with Scylla - our driver automatically establishes 1 connection per shard, so it was immediately able to reach >150kops locally, while `cassandra-cpp` used for Scylla was capped at half of that (since it wouldn't use the whole potential by not connecting to all shards).

P.S. I patched your `scylla-driver` branch to allow configuring the number of connections per host with:
```diff
diff --git a/src/session.rs b/src/session.rs
index b7026f3..756db78 100644
--- a/src/session.rs
+++ b/src/session.rs
@@ -3,11 +3,14 @@ use itertools::Itertools;
 use scylla::transport::errors::{NewSessionError, QueryError};
 use scylla::{Session, SessionBuilder};
 use std::process::exit;
+use scylla::transport::session::PoolSize;
+use std::num::NonZeroUsize;
 
 /// Configures connection to Cassandra.
 pub async fn connect(conf: &RunCommand) -> Result<Session, NewSessionError> {
     SessionBuilder::new()
         .known_nodes(&conf.addresses)
+        .pool_size(PoolSize::PerHost(NonZeroUsize::new(conf.connections).unwrap()))
         .build()
         .await
 }
```
you're welcome to steal the change and put it on your branch, to make it more configurable :)
